### PR TITLE
Get Forsyth-Edwards Notation : GetFen()

### PR DESCRIPTION
**Pull request by MelleKoning**
Requested by jjosh on Discord, addition of a GetFen() to create a fen from current position.

### CODE REVIEW

**Comment by MelleKoning** [comment]
The test reveals the GetFen() for black positions does not function properly. Does anybody know what is different in the ChessBoard when it's black to move and thus why GetFen() gives strange results?
Please see position_test.cc

**Comment by MelleKoning** [comment]
added a command to get the current position from the engine.

When no search is done yet, the latest known fen (startpos) is returned.
After search, the fen of current position is returned.

Example:


```
position startpos moves e2e4 e7e5
getfen
rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 0
go nodes 2
bestmove g1f3 ponder g8f6
getfen
rnbqkbnr/pppp1ppp/8/4p3/4P3/8/PPPP1PPP/RNBQKBNR w KQkq - 0 2
```

**Comment by MelleKoning** [comment]
From discord: author of Nibbler, a GUI for Lc0, wrote a fen parser.

Plea for merge: It can be tricky to get the fen writer reader right to also include enpassent, maybe it helps GUI writers in general to not have to reinvent the wheel. Having a fen reader will therefore be beneficial for lc0 gui writers. It just saves them for having to write that on their own when the engine alteady supports it via uci commands.

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: It's better to declare variables at the point of the first use rather than in the beginning of the function.

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: With `GetWhiteBoard()`, no need to flip it and have this variable.

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: I think this logic should be moved to another function. Other a free-standing one or member function of `ChessBoard` class. Something like:

```cpp
char GetPieceAt(const ChessBoard& board, int row, int col) {
  BoardSquare sq(row, col);
  if (!(ours() | theirs()).get(sq)) return '\0';
  if (board.our_king() == sq) return 'K';
  if (board.their_king() == sq) return 'k';
  char res = 'N';
  if (board.bishops().get(sq)) res = 'B';
  else if (board.queens().get(sq)) res = 'Q';
  else if (board.rooks().get(sq)) res = 'R';
  return flipped() ? std::tolower(res) : res;
}

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: From reading the code, it seems capitals are for white.

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: `const ChessBoard& side_to_move_board = GetBoard();`

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: Actually as rank is always `3` for black en passants and `6` for white, no need for side_to_move_board at all.

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: A bit hacky, but saves some typing. :) And no need for loop.
```
auto sq = *board.en_pasant().begin();
result += std::string(1, 'a' + sq.col()) + IsBlackToMove() ? "3" : "6";
```

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]:
```cpp
If (board.en_passant().empty()) {
 result += "-";
} else {
...
```

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc]: Can call `board.castlings().as_string()` here without any temporary variables.

File: src/chess/uciloop.cc
Comment (mooskagh) [inline src/chess/uciloop.cc]: There's no uci command "get" (I've just checked stockfish doesn't have it either).
If stockfish supports it in some way, it's better to be consistent.
If not, maybe we don't need it either.

In any case, I don't think that adding non-standard commands in `UciLoop` is a good idea.
Instead, we should probably add `virtual bool CmdNonStandardCommand(const std::string& command);` which uci loop should call in such cases, and `EngineLoop::CmdNonStandardCommand` should just have `if (command == "blah") DoThings(); `

File: src/engine.cc
Comment (mooskagh) [inline src/engine.cc]: Throw exception (from `utils/exception.h`) instead.

File: src/engine.h
Comment (mooskagh) [inline src/engine.h]: Better `GetCurrentPositionFen()`.

File: src/chess/position_test.cc
Comment (MelleKoning) [reply src/chess/position_test.cc]: Installed clang-format and ran it.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Removed

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Thanks, introduced GetWhiteBoard();

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Moved enpassant to where it is used

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Removed fenflipped

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Yes, just initializing as empty with dash(-) and overwriting when there is an en_passent space

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: fixed

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Adjusted the comment

Comment (MelleKoning) [reply src/chess/position.cc]: Adjusted the comment

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: I don't think it is stored at rank three and six but on 0 and 7; when I asked about en_passent Tilps in discord pointed me to the code here which I simply re-used here: https://github.com/Tilps/lc0/blob/rescore_tb/src/selfplay/loop.cc#L282

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Have changed the for loop to a hacky line but at least kept using the provided as_string()

File: src/chess/uciloop.cc
Comment (MelleKoning) [reply src/chess/uciloop.cc]: There is no uci command indeed. Now that lc0 is champion we could set our own standards.

Just kidding ;) 
Introduced a CmdNonStandardUciCommand method and moved code up to EngineLoop. However, there is a command-parser in UciLoop which I can't circumvent easily, so "get","fen" is still there...

File: src/engine.cc
Comment (MelleKoning) [reply src/engine.cc]: Now using throw exception

File: src/engine.h
Comment (MelleKoning) [reply src/engine.h]: Renamed.

File: src/chess/uciloop.cc
Comment (mooskagh) [inline src/chess/uciloop.cc]: My idea was that `CmdNonStandardUciCommand` returns `true` if it managed to process command and `false` if not, so here it would still be:
```cpp
else if (!CmdNonStandardUciCommand(command, params)) {
  throw Exception("Unknown command: " + command);
}
```

And in `UciLoop::CmdNonStandardUciCommand` just `return false`

File: src/engine.cc
Comment (mooskagh) [inline src/engine.cc]: Why not just a `getfen`? Is it so in some other engine?
Also, maybe just pass unparsed command to `CmdNonStandardUciCommand` (not command and params), so that it can parse it as it wishes. That way we can even compare with `"get fen"` (because currently it would work with `get asd asd fen sdjksd sdfkls ldkjfds` but won't with `get asd fen sdjkfl sdjklsdjk`, which is not very nice).

Comment (MelleKoning) [reply src/engine.cc]: First I started out with 'getfen' but I did not like it as it is not English, and does not seem to match the other command/parameter pairs. I future we might want to support to get other stuff. 'get fen' is just a starting point. For analyses purposes we could extend further with other non-uci commands. Not sure what though, yet but for example 'get epd' if chess960 would be supported. It's not really part of the scope and when something new is introduced we can change this later.

File: src/chess/uciloop.cc
Comment (MelleKoning) [reply src/chess/uciloop.cc]: Found a way to move out 'get fen' from the uci commands. Having the sole 'line string' loses the parsing of command and tokens, but I guess it's good for now.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Yes, that makes the algorithm much better. Adjusted the code.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Done

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Altered.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Yes, good suggestion! Had to switch the 2 and 5, black can take on row 2, white on row 5 :)

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Adjusted.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Fixed.

File: src/chess/position.h
Comment (MelleKoning) [reply src/chess/position.h]: Have clang-format installed as a plug-in to visual studio. Must say that the default behavior is not always ideal, that is, I find the lines a little bit too short. For example: I do not use the clang-format on the position_test.cc file because it breaks up the fen-strings!

File: src/chess/uciloop.h
Comment (MelleKoning) [reply src/chess/uciloop.h]: Adjusted.

File: src/chess/uciloop.cc
Comment (MelleKoning) [reply src/chess/uciloop.cc]: Had to re-wire a few things to ensure the {"get","fen"} was moved out of the uciloop.cc file. Now indeed the typed in line will be forwarded to the engine class without command/token parsing.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Used the '?' during creation to see if I got all different pieces covered. Now that it works indeed no longer needed; changed to initialize with zero-char now.

File: src/chess/uciloop.cc
Comment (mooskagh) [inline src/chess/uciloop.cc]: Does it work this way?.. We should get there only when command is empty, but in this case CmdNonStandardUciCommand() will have nothing to match against.

I suggest we either call `CmdNonStandardUciCommand()` from within DispatchCommand() (and in this case we need to add "get" command into `kKnownCommands` like it's done for non-standard "xyzzy" command), or

call "TryHandleNonStandardCommand()" before calling ParseCommand() (and skip parsing if it's successful). That doesn't look very nice though.

File: src/chess/uciloop.h
Comment (mooskagh) [inline src/chess/uciloop.h]: I think default behaviour should be `return false;` rather than throwing exception?

File: src/engine.cc
Comment (mooskagh) [inline src/engine.cc]: Actually we need `position startpos|fen` I think, and not go nodes.
I think error message like "No game in progress", "Game not started" etc would be more appropriate.

File: src/engine.cc
Comment (mooskagh) [inline src/engine.cc]: Instead of duplicating the function, maybe it makes sense to create `utils/misc.h` and move it there? (also use it in uciloop.cc). Something like:

```cpp
template<class T, U>
bool ContainsKey(const T& container, const U& key) {
   return container.find(key) != container.end();
}

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Sorry, I don't get it. I moved it close to top of the file, still inside namespace lczero already.. I can't put the function outside (on top of) namespace lczero as it has a reference to ChessBoard. 
FYI: I'm not all too familiar with C++ 11 so please enlighten me :)

File: src/chess/uciloop.h
Comment (MelleKoning) [reply src/chess/uciloop.h]: Done

File: src/engine.cc
Comment (MelleKoning) [reply src/engine.cc]: Hm, the way the fen is looked up, the position should already have been processed through the engine. For some reason this set of commands does not work:

> position fen 8/2p5/3p4/KP5r/1R3p1k/8/4P1P1/8 w - - 1 1
> get fen
> error Need a 'go nodes..' before returning fen position.

but this does:

 
> position fen 8/2p5/3p4/KP5r/1R3p1k/8/4P1P1/8 w - - 1 1
> go nodes 1
> ...
> get fen
> 8/2p5/3p4/KP5r/1R3p1k/8/4P1P1/8 w - - 1 0

File: src/engine.cc
Comment (MelleKoning) [reply src/engine.cc]: Good catch; it was dead code. Is now removed.

File: src/chess/position.cc
Comment (MelleKoning) [reply src/chess/position.cc]: Ok, found out what anonymous namespace is and moved it there

**Comment by Naphthalin** [comment]
What is the current stance of this PR? If this isn't implemented yet but useful, we should decide on merging it after an update.

**Comment by MelleKoning** [comment]
What update would be needed to enable merge?

**Comment by Naphthalin** [comment]
it shows 3 merge conflicts, that's what I referred to as "update". But I think before doing the work, the decision whether this will be merged should be made.

File: src/engine.h
Comment (mooskagh) [inline src/engine.h:137]: This is not needed anymore.

**Comment by mooskagh** [comment]
This was/is almost ready, I believe it never got back after a round of comments.
After rebasing and checking that it works fine with FRC, I think it's good to be merged.

**Comment by liamlim** [comment]
What is needed for this PR to get merged? Is some testing needed? Maybe I could help but I think it would be good to know from maintainers what is the state of this PR. It was reviewed many times and now it's rebased on current master

**Comment by Naphthalin** [comment]
We are currently in the process of collecting which PRs to merge into `v0.26.0`, so right now PRs are getting selected with having that in mind.

**Comment by Naphthalin** [comment]
@liamlim it seems to me that there are a few comments by @mooskagh which weren't addressed yet; other than that I don't see anything holding this PR back.

**Comment by liamlim** [comment]
What exactly is necessary to test about FRC for this PR to address comments by mooskagh? I think that the only think that makes FRC different from regular chess are castling rights and these work just fine in my opinion.

**Comment by MelleKoning** [comment]
This PR is long due. It works any updates could

**Comment by liamlim** [comment]
I think that all comments in this PR has been addressed and it works just fine. Is there any additional work needed for this to get merged?

**Comment by borg323** [comment]
A new review is needed, since I did extensive modifications.

**Comment by Naphthalin** [comment]
@mooskagh any chance you can take a more thorough look here? The inner function `Position::GetFEN()` is likely needed for transposition related things, especially for updating the time manager to account for transpositions near root.

File: src/chess/uciloop.cc
Comment (mooskagh) [reply src/chess/uciloop.cc:215]: Ok, after one more thought... It's becoming too complicated, so I take back my suggestion from 2 years ago, sorry to change my mind.

Now I believe it's cleaner just to add yet another command (`fen`) to `kKnownCommands` and add `CmdFen()` to `UciLoop`.

File: src/engine.cc
Comment (mooskagh) [inline src/engine.cc:179]: I think it would be cleaner to split this function into two. How about that:

1. Make `current_position_` non `std::optional` (initialize it to `ChessBoard::kStartposFen, {}` on construction in in new_game`.
2. Remove `else` branch in `if (current_position_)` in `EngineController::Go`.
3. Have function `Position FoldPosition(const CurrentPosition& pos)` which would do what this branch does (no need to put it in a class). (or some better name than `FoldPosition` but it's good to have a name that implies that it's not a O(1) operation... `BuildPosition()`? `ApplyPositionMoves()`?)
4. Call `FoldPosition(current_position_).GetFen()` instead of `GetCurrentPositionFen()`.
5. (optional) rename `CurrentPosition` (struct name, not the variable) to something better. :)

File: src/chess/position.cc
Comment (mooskagh) [inline src/chess/position.cc:164]: This function doesn't really access (or has to access) any private member variables, consider making it a free-standing function rather than a member. <https://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197>, <https://flylib.com/books/en/2.887.1.37/1/>

File: src/engine.h
Comment (mooskagh) [reply src/engine.h:132]: This is needed again...
Sorry for making to do and undo extra work, I've just took a look what it would take to add this "NonStandardUciCommand" function cleanly and it really doesn't seem to be that clean.

**Review by mooskagh — CHANGES_REQUESTED** [review]

File: src/chess/board.h
Comment (borg323) [reply src/chess/board.h:142]: Seems to work fine in a few tests (mostly lichess and cutechess). It can be changed, but is it worth it? E.g. start position 508 would be `rbqnknbr/pppppppp/8/8/8/8/PPPPPPPP/RBQNKNBR w KQkq - 0 1` if we also check the king's position rather than `rbqnknbr/pppppppp/8/8/8/8/PPPPPPPP/RBQNKNBR w AHah - 0 1` and getting this right is not trivial. (Also putting the later form in the lichess board editor loses black's castling rights).
So I'd rather leave it like that until we get actual issues reported.

File: src/engine.cc
Comment (borg323) [reply src/engine.cc]: Note that now the position is never taken from the tree - this is implied in the above and makes in practice, as a `position` followed by `fen` would give the fen for the tree head with the previous version instead of the just set position.

**Comment by borg323** [comment]
With the last revision I changed the command to a simple `fen`, if anyone objects now is the time to speak up.

**Comment by MelleKoning** [comment]
Nice work borg!

**Review by mooskagh — APPROVED** [review]

**Pull request merged**
